src/dataset.cpp

```diff
@@ -514,8 +515,81 @@ Dataset concatenate(const Dataset &d1, const Dataset &d2, const Dim dim) {
+      d.get<const Coord::ComponentInfo>()[0].get<const Coord::Position>();
+  // TODO Need a better mechanism to identify source and sample.
+  const auto &sourcePos = compPos[0];
+  const auto &samplePos = compPos[1];
+  const double l1 = (sourcePos - samplePos).norm();
+
+  const auto &dims = d.contains(Coord::Position{})
+                         ? d(Coord::Position{}).dimensions()
+                         : d(Coord::DetectorGrouping{}).dimensions();
+  Variable conversionFactor(Data::Value{}, dims);
+
+  MDZipView<const Coord::Position> specPos(d);
+  const auto factor = [&](const auto &item) {
+    const auto &pos = item.template get<Coord::Position>();
+    double l_total = l1 + (samplePos - pos).norm();
+    // TODO Up to physical constants.
+    return l_total * l_total;
+  };
+  // TODO Must also update unit of conversionFactor.
+  std::transform(specPos.begin(), specPos.end(),
+                 conversionFactor.get<Data::Value>().begin(), factor);
+
+  // 2. Transform variables
+  Dataset converted;
+  for (const auto &var : d) {
+    auto varDims = var.dimensions();
+    if (varDims.contains(Dim::Tof))
+      varDims.relabel(varDims.index(Dim::Tof), Dim::Energy);
+    if (var.tag() == Coord::Tof{}) {
+      // TODO Need to extend the broadcasting capabilities to broadcast to the
+      // union of dimensions of both operands in a binary operation.
+      auto dims = conversionFactor.dimensions();
+      for (const Dim dim : varDims.labels())
+        if (!dims.contains(dim))
+          dims.addInner(dim, varDims[dim]);
+      // TODO Should have a broadcasting assign method?
+      auto energy = makeVariable<Coord::Energy>(dims, dims.volume(), 1.0);
+      energy *= conversionFactor;
+      // TODO Change this to /= when implemented.
+      // The reshape is just to remap the dimension label, should probably do
+      // this differently.
+      energy *= (var * var).reshape(varDims);
+      converted.insert(energy);
+    } else if (var.tag() == Data::Events{}) {
+      throw std::runtime_error(
+          "TODO Converting units of event data not implemented yet.");
+    } else {
+      // TODO Changing Dim::Tof to Dim::Energy. Currently this is not possible
+      // without making a copy of the data, which is inefficient. If we cannot
+      // move the dimensions outside the cow_ptr in Variable, another option
+      // would be to support in-place modification. Probably a better solution
+      // would be to decouple the pointer holding VariableConcept from the COW
+      // mechanism, i.e., to COW inside VariableConcept to hold the data
+      // array?
+      // TODO Also need to check here if variable contains count/bin_width,
+      // should fail then?
+      converted.insert(var.reshape(varDims));
+    }
+  }
+
+  return converted;
+}
+} // namespace tof
+} // namespace neutron
+
 Dataset convert(const Dataset &d, const Dim from, const Dim to) {
-  static_cast<void>(to);
+  if ((from == Dim::Tof) && (to == Dim::Energy))
```

Comment:
I still have to figure out how to do this. Mantid currently always converts via TOF (unless there are special cases enabled), i.e., when you request conversion from A to B it does A -> TOF -> B. Ideally I would like to avoid this, but it is not clear to me yet whether we can do that with help of the compiler, or whether we would have to hand-code all possible conversions.

The specialization of `convert`, or how it dispatches to the actual conversion method will depend on that... so I think we have to postpone this until the problem is better understood and refactor then.